### DISCUSSION TITLE
Firebase authentication and authorization

### DISCUSSION THREAD

**Discussion by defrimhasani** [quarkusio/quarkus#28482] [Q&A]
Hi everyone,

Wanted to ask if anyone has tried adding firebase authn & authz in a qurakus app? If yes, can you please point me to any valid resource?

Thanks,
Defrim

**Comment by sberyozkin** [comment]
Hi @defrimhasani I'll try to look at https://github.com/smallrye/smallrye-jwt/issues/382 as soon as I can

**Comment by sberyozkin** [comment]
@defrimhasani I've finally had a chance to look a bit deeper into this issue. The answer has been found in [this stacktrace post](https://stackoverflow.com/a/71096804/1133928).

Specifically, all you need to do is to configure smallrye-jwt to point to the JWK store format:

https://www.googleapis.com/service_accounts/v1/jwk/securetoken@system.gserviceaccount.com.

If you have a Firebase ID token with a `kid` header then it will work. Can you try it ?

**Reply by md-consulting** [reply]
@sberyozkin Thank you for providing this answer. Curiously I found that after including Google API Gateway in the communication between Firebase and quarkus back-end (Google Cloud Run), this no longer seems to work.

Could you please check [my question on StackOverflow](https://stackoverflow.com/questions/78586826/how-to-retrieve-firebase-jwt-key-for-quarkus-native-app-deployed-on-google-clo)?

(Everything worked before adding API Gateway the way you suggested it)

Thank you very much in advance and best regards

**Reply by sberyozkin** [reply]
@md-consulting Np, if I understand it correctly, the [Gateway authenticates the token itself](https://cloud.google.com/api-gateway/docs/authenticating-users-firebase#configuring_apig_to_support_client_authentication) and then forwards the authentication results to the endpoint. Unless I'm mistaken, in this case you should register a custom factory accepting an already authenticated token: 
https://quarkus.io/guides/security-jwt#custom-factories

You should probably configure smallrye-jwt to expect `X-Apigateway-Api-Userinfo` to keep a token

Can you try it please ? 
Let me add the same comment in the SO question

**Reply by md-consulting** [reply]
Hi @sberyozkin,
That worked like a charm. I set `mp.jwt.token.header=X-Apigateway-Api-Userinfo` and created a custom JWTCallerPrincipalFactory. However I did not need to split the token before decoding (`...token.split(*\\.*)[1]` threw an error because the length of the array was 1), so I simply decoded the token without splitting.
Thank you very much!

**Reply by sberyozkin** [reply]
Sounds good, thanks @md-consulting
